contracts/utils/VaultQueue.sol

```diff
@@ -0,0 +1,223 @@
+
+    function transfer(address vault) external onlyOwner nonReentrant {
+        uint256 withdrawals = IRibbonVault(vault).withdrawals(address(this));
+        if (withdrawals > 0) {
+            if (vault == stethVault) {
+                IRibbonVault(vault).completeWithdraw(0);
+                disburse(vault);
+            } else {
+                IRibbonVault(vault).completeWithdraw();
+                disburse(vault);
+            }
+        }
+    }
+
+    function queueTransfer(
+        address srcVault,
+        address dstVault,
+        address depositContract,
+        TransferAction transferAction,
+        uint256 amount
+    ) external nonReentrant {
+        require(
+            !hasWithdrawal(srcVault, msg.sender),
+            "Withdraw already submitted"
+        );
+        require(qTransfer[srcVault].length < 256, "Transfer queue full");
+
+        if (transferAction == TransferAction.INTERVAULT) {
+            require(depositContract != address(0), "Intervault !address");
+        } else if (transferAction == TransferAction.WITHDRAW) {
+            // depositContract is not used on withdraw, so let's set it to msg.sender
+            require(
+                depositContract == msg.sender,
+                "On withdraw, depositContract must be msg.sender"
+            );
+        }
+
+        IRibbonVault(srcVault).transferFrom(msg.sender, address(this), amount);
+
+        IRibbonVault(srcVault).initiateWithdraw(amount);
+
+        qTransfer[srcVault].push(
+            Transfer(
+                msg.sender,
+                srcVault,
+                dstVault,
+                depositContract,
+                uint32(block.timestamp),
+                amount,
+                transferAction
+            )
+        );
+        totalAmount[srcVault] += amount;
+    }
+
+    function hasWithdrawal(address vault, address user)
+        public
+        view
+        returns (bool)
+    {
+        for (uint256 j = 0; j < qTransfer[vault].length; j++) {
+            if (user == qTransfer[vault][j].creditor) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    function pop(Transfer[] storage array) private returns (Transfer memory) {
+        Transfer memory item = array[array.length - 1];
+        array.pop();
+        return item;
+    }
+
+    function reset(address vault) external onlyOwner {
+        delete qTransfer[vault];
+        totalAmount[vault] = 0;
+    }
+
+    // ETH and ERC20s are not kept in this contract unless explicitly sent.
+    // This contract intermediately holds tokens only during transfer() and only within the transaction.
+    function rescueETH(uint256 amount) external onlyOwner {
+        payable(msg.sender).transfer(amount);
+    }
+
+    function rescue(address asset, uint256 amount) external onlyOwner {
+        IERC20(asset).approve(address(this), amount);
+        IERC20(asset).transfer(msg.sender, amount);
+    }
+
+    receive() external payable {}
```

Comment:
Looping back to what we talked about before, let's add a check to verify that it's coming from the vault

```solidity
require(msg.sender == wethVault || msg.sender == stethVault);
```

Reply:
done